### PR TITLE
BatchSplit

### PR DESCRIPTION
**Pull request by hSterz**
Add new composition block BatchSplit

### CODE REVIEW

File: src/transformers/adapters/layer.py [hunk 1]
```diff
@@ -329,6 +346,59 @@ def adapter_parallel(self, adapter_setup: Parallel, hidden_states, input_tensor,
...
+                children_hidden.append(child)
+            # Case 5: nesting other composition blocks is invalid
+            elif isinstance(adapter_block, AdapterCompositionBlock):
+                raise ValueError(
+                    "Invalid adapter setup. Cannot nest {} in {}".format(
+                        adapter_block.__class__.__name__, adapter_setup.__class__.__name__
+                    )
+                )
+            # Case X: No adapter which is part of this module -> ignore
+
```
Comment (calpt) [inline src/transformers/adapters/layer.py:398]:
```suggestion
else:
    children_hidden.append(hidden_states[batch_idx])

```

**Review by calpt — COMMENTED** [review]
Looks good overall

File: tests/test_adapter_heads.py [hunk 2]
```diff
@@ -146,6 +147,16 @@ def test_load_full_model(self):
...
 
+    def test_batch_split_head(self):
+        model = AutoModelWithHeads.from_config(self.config())
+        model.add_classification_head("a")
+        model.add_classification_head("b")
+        model.active_head = BatchSplit("a", "b", batch_sizes=[1, 2])
+        in_data = self.get_input_samples((3, 128), config=model.config)
+
+        out = model(in_data)
+        self.assertEqual(2, len(out))
```
Comment (calpt) [inline tests/test_adapter_heads.py:158]:
```suggestion
        self.assertEqual(2, len(out))
        self.assertEqual((1, 2), out[0][0].shape)
        self.assertEqual((2, 2), out[1][0].shape)
```

**Review by calpt — COMMENTED** [review]
Looks nice!

Could you also undo the file moves in the examples? Probably got messed up with a merge

File: src/transformers/adapters/heads.py [hunk 3]
```diff
@@ -529,32 +531,54 @@ def forward_head(
...
 
+        def _get_head_input(outputs, cls_out, batch):
+            # TODO-AH check possible edge cases here
+            if isinstance(outputs, ModelOutput):
+                inputs = {}
+                for key, base_output in outputs.items():
+                    if isinstance(base_output, tuple):
+                        inputs[key] = base_output[batch_idx[0] : batch_idx[-1]]
+                    else:
+                        inputs[key] = base_output[batch]
```
[the remark this answers is not in the document]
Comment (hSterz) [reply src/transformers/adapters/heads.py]: That is true  :sweat_smile: I cleaned that up

File: src/transformers/adapters/heads.py [hunk 4]
```diff
@@ -529,32 +531,54 @@ def forward_head(
...
+        if isinstance(self.active_head, BatchSplit):
+            if sum(self.active_head.batch_sizes) != all_outputs[0].size()[0]:
+                raise ValueError(
+                    "The specified batch sizes {} do not match the actual batch size {}".format(
+                        self.active_head.batch_sizes, all_outputs[0].size()[0]
+                    )
+                )
+            head_outputs = []
+            for i, head in enumerate(self.active_head):
+                head_module = self.heads[head]
```
[the remark this answers is not in the document]
Comment (hSterz) [reply src/transformers/adapters/heads.py:585]: That is a good point. I added that the active head is adapted from the active adapter setup accordingly.

File: src/transformers/adapters/heads.py [hunk 5]
```diff
@@ -483,7 +483,14 @@ def set_active_adapters(
...
             elif isinstance(final_block, BatchSplit):
-                self.active_head = final_block
+                blocks = [
+                    block.last() if isinstance(block, AdapterCompositionBlock) else block for block in final_block
+                ]
+                head_setup = BatchSplit(*blocks, batch_sizes=final_block.batch_sizes)
+                if all(head in self.heads for head in head_setup):
+                    self.active_head = head_setup
+                else:
+                    raise ValueError("Missing head for the given BatchSplit setup expected heads {}".format(blocks))
```
Comment (calpt) [inline src/transformers/adapters/heads.py:493]:
```suggestion
                    raise ValueError("Missing at least one head for the given BatchSplit setup. Expected heads: {}".format(blocks))
```

File: src/transformers/adapters/heads.py [hunk 6]
```diff
@@ -483,7 +483,14 @@ def set_active_adapters(
             elif isinstance(final_block, Parallel):
                 self.active_head = [a if isinstance(a, str) else a.last for a in final_block.children]
             elif isinstance(final_block, BatchSplit):
-                self.active_head = final_block
+                blocks = [
+                    block.last() if isinstance(block, AdapterCompositionBlock) else block for block in final_block
```
Comment (calpt) [inline src/transformers/adapters/heads.py:487]:
```suggestion
                    block.last if isinstance(block, AdapterCompositionBlock) else block for block in final_block
```

File: src/transformers/adapters/heads.py [hunk 7]
```diff
@@ -483,7 +483,14 @@ def set_active_adapters(
             elif isinstance(final_block, Parallel):
                 self.active_head = [a if isinstance(a, str) else a.last for a in final_block.children]
             elif isinstance(final_block, BatchSplit):
-                self.active_head = final_block
+                blocks = [
```
Comment (calpt) [inline src/transformers/adapters/heads.py:486]:
```suggestion
                # Convert BatchSplit of adapters to a BatchSplit of heads.
                blocks = [
```

**Review by calpt — APPROVED** [review]
Looks nice! 🚀

File: src/transformers/adapters/heads.py (hunk 6 above)
Comment (hSterz) [reply src/transformers/adapters/heads.py:488]: This needs to be a method call otherwise it does not work

Comment (calpt) [reply src/transformers/adapters/heads.py:488]: Oh, sorry, thought it was a property based on `a.last` four lines above (which in that case also should not work).

**Pull request merged**